juju/juju
# [JUJU-2831] File notify watcher

The following is a file notify watcher, that wraps the inotify syscall
on linux. This is a much more efficient implementation than the wrench
package, as we're leaning onto the kernel to emit a change if there is
any change to a folder.

We will watch a given folder for file creation/deletion and then perform
a block/unblock on the change stream.

The premise behind this is to be able to control the changestream
without using the database. This is currently just a open/close valve,
but I don't see why we can't use it for throttling the changestream to
certain events or the number of events.

There is a lot of plumbing because of the basic boilerplate of workers,
but I do believe this should be resilient enough for what we want.

## Checklist


## QA steps

```sh
$ juju bootstrap lxd test --build-agent
```

To see a working example, this is quite convoluted:

  1. Apply the following diff at the bottom of this post.
  2. `$ juju bootstrap lxd test --build-agent`
  3. `make repl`
  4. Run `.open controller` in the repl to open the database table
  5. Insert a namespace, this is to satisfy the constraints `INSERT INTO change_log_namespace VALUES (1, 'foo');`
  6. Watch `juju debug-log -m controller` from one terminal
  7. Insert a change log item `INSERT INTO change_log (edit_type_id, namespace_id, changed_uuid) VALUES (2, 1, 'uuid1');`
 
You should now see the following logs:

```
machine-0: 17:26:23 ERROR juju.worker.changestream change event: {2 2 foo uuid1 2023-02-09 17:26:23.227}
machine-0: 17:26:23 ERROR juju.worker.lease lease manager got changes: {2 2 foo uuid1 2023-02-09 17:26:23.227}
```

  8. Now `lxc exec <juju container> - bash`
  9. `touch /var/lib/juju/locks/controller.db.lock`
 
You should now see the following logs:

```
machine-0: 12:08:52 INFO juju.worker.changestream Change stream has been blocked
```

  10. Insert a load more change log items
  11. `rm /var/lib/juju/locks/controller.db.lock`

You should now see only 1 item for the same entity, as we're coalescing:

```
machine-0: 12:09:10 ERROR juju.worker.lease lease manager got changes: {8 2 foo uuid1 2023-02-15T12:09:00.784Z}
```


------

Diff to apply to see file notify watcher working:

```diff
diff --git a/cmd/jujud/agent/machine/manifolds.go b/cmd/jujud/agent/machine/manifolds.go
index dfc26d367c..1e7fdeae5b 100644
--- a/cmd/jujud/agent/machine/manifolds.go
+++ b/cmd/jujud/agent/machine/manifolds.go
@@ -733,6 +733,7 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
                        AgentName:            agentName,
                        ClockName:            clockName,
                        DBAccessorName:       dbAccessorName,
+                       ChangeStreamName:     changeStreamName,
                        Logger:               loggo.GetLogger("juju.worker.lease"),
                        LogDir:               agentConfig.LogDir(),
                        PrometheusRegisterer: config.PrometheusRegisterer,
diff --git a/worker/lease/manifold/manifold.go b/worker/lease/manifold/manifold.go
index a5e6420f60..f00e1c1e6a 100644
--- a/worker/lease/manifold/manifold.go
+++ b/worker/lease/manifold/manifold.go
@@ -20,6 +20,7 @@ import (

        "github.com/juju/juju/agent"
        corelease "github.com/juju/juju/core/lease"
+       "github.com/juju/juju/worker/changestream"
        "github.com/juju/juju/worker/common"
        "github.com/juju/juju/worker/dbaccessor"
        "github.com/juju/juju/worker/lease"
@@ -35,9 +36,10 @@ const (
 // ManifoldConfig holds the resources needed to start the lease
 // manager in a dependency engine.
 type ManifoldConfig struct {
-       AgentName      string
-       ClockName      string
-       DBAccessorName string
+       AgentName        string
+       ClockName        string
+       DBAccessorName   string
+       ChangeStreamName string

        Logger               lease.Logger
        LogDir               string
@@ -102,6 +104,24 @@ func (s *manifoldState) start(context dependency.Context) (worker.Worker, error)
                return nil, errors.Trace(err)
        }

+       var changeStream changestream.ChangeStream
+       if err := context.Get(s.config.ChangeStreamName, &changeStream); err != nil {
+               return nil, errors.Trace(err)
+       }
+
+       changes, err := changeStream.Changes("controller")
+       if err != nil {
+               return nil, errors.Trace(err)
+       }
+       go func() {
+               for {
+                       select {
+                       case change := <-changes:
+                               s.config.Logger.Errorf("lease manager got changes: %v", change)
+                       }
+               }
+       }()
+
        s.store = s.config.NewStore(lease.StoreConfig{
                DB:     db,
                Logger: s.config.Logger,
@@ -146,6 +166,7 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
                        config.AgentName,
                        config.ClockName,
                        config.DBAccessorName,
+                       config.ChangeStreamName,
                },
                Start:  s.start,

```
